mailpoet/lib/Automation/Integrations/WooCommerce/Triggers/OrderStatusChangedTrigger.php

```diff
@@ -0,0 +1,102 @@
+<?php declare(strict_types = 1);
+
+namespace MailPoet\Automation\Integrations\WooCommerce\Triggers;
+
+use MailPoet\Automation\Engine\Data\StepRunArgs;
+use MailPoet\Automation\Engine\Data\StepValidationArgs;
+use MailPoet\Automation\Engine\Data\Subject;
+use MailPoet\Automation\Engine\Hooks;
+use MailPoet\Automation\Engine\Integration\Trigger;
+use MailPoet\Automation\Integrations\WooCommerce\Payloads\OrderStatusChangePayload;
+use MailPoet\Automation\Integrations\WooCommerce\Subjects\CustomerSubject;
+use MailPoet\Automation\Integrations\WooCommerce\Subjects\OrderStatusChangeSubject;
+use MailPoet\Automation\Integrations\WooCommerce\Subjects\OrderSubject;
+use MailPoet\Validator\Builder;
+use MailPoet\Validator\Schema\ObjectSchema;
+use MailPoet\WooCommerce;
+use MailPoet\WP\Functions;
+
+class OrderStatusChangedTrigger implements Trigger {
+
+  /** @var Functions */
+  private $wp;
+
+  /** @var WooCommerce\Helper */
+  private $woocommerceHelper;
+
+  public function __construct(
+    Functions $wp,
+    WooCommerce\Helper $woocommerceHelper
+  ) {
+    $this->wp = $wp;
+    $this->woocommerceHelper = $woocommerceHelper;
+  }
+
+  public function getKey(): string {
+    return 'woocommerce:order-status-changed';
+  }
+
+  public function getName(): string {
+    return __('Order status changed', 'mailpoet');
+  }
+
+  public function getArgsSchema(): ObjectSchema {
+    return Builder::object([
+      'from' => Builder::string()->required()->default('any'),
+      'to' => Builder::string()->required()->default('wc-completed'),
+    ]);
+  }
+
+  public function getSubjectKeys(): array {
+    return [
+      OrderSubject::KEY,
+      OrderStatusChangeSubject::KEY,
+      CustomerSubject::KEY,
+    ];
+  }
+
+  public function validate(StepValidationArgs $args): void {
+  }
+
+  public function registerHooks(): void {
+    $this->wp->addAction(
+      'woocommerce_order_status_changed',
+      [
+        $this,
+        'handle',
+      ],
+      10,
+      3
+    );
+  }
+
+  public function handle(int $orderId, string $oldStatus, string $newStatus): void {
+    $order = $this->woocommerceHelper->wcGetOrder($orderId);
+    if (!$order instanceof \WC_Order) {
+      return;
+    }
+
+    $this->wp->doAction(Hooks::TRIGGER, $this, [
+      new Subject(OrderStatusChangeSubject::KEY, ['from' => $oldStatus, 'to' => $newStatus]),
+      new Subject(OrderSubject::KEY, ['order_id' => $order->get_id()]),
+      new Subject(CustomerSubject::KEY, ['customer_id' => $order->get_customer_id()]),
+    ]);
+  }
+
+  public function isTriggeredBy(StepRunArgs $args): bool {
+    $orderPayload = $args->getSinglePayloadByClass(OrderStatusChangePayload::class);
```

Comment:
The above method will throw an exception if there is no single payload, so the check below is not needed. Unless you want to catch the exception and return `false` but I guess that's not necessary.

Reply:
[1296e82](https://github.com/mailpoet/mailpoet/pull/4794/commits/1296e826c8167241e1615e88eba9e37d238de9b2)